### ISSUE TITLE
Trying to understand parsing of document with Form XObject

### ISSUE THREAD

**Issue by drcdr** [qpdf/qpdf#436]
I am trying to understand the behavior of the parsing of a PDF file with Form XObjects, created with pdfTex.  The file I'm using is [this one](https://arxiv.org/abs/1803.07133).

I'm basing my parsing of the Form XObjects on what I see using `qpdf --json`.    Here are the key parts of the json dump, for page 1.  [My comments in brackets]
```
"12 0 R": {  [Root]
    "/Count": 6, 
    "/Kids": ["3 0 R","66 0 R","91 0 R","123 0 R","133 0 R","138 0 R"],
    "/Parent": "179 0 R","/Type": "/Pages"
}
"3 0 R": {  [Kid 1]
     "/Contents": ["9 0 R","187 0 R"],
     "/MediaBox": [0,0,612,792], 
     "/Parent": "12 0 R", 
     "/Resources": "8 0 R",
     "/Type":"/Page"
},
"8 0 R": {  [ Note: Even through there is a font here ...  there is an /XObject as well. ]
     "/Font": {"/arXivStAmP": "188 0 R"},
     "/ProcSet": ["/PDF"],
     "/XObject": {"/Im4": "7 0 R"}
},
"7 0 R": {  [ I can't seem to find this object with QPDFObjectHandle functions ]
     "/BBox": [0, 0, 612, 792], 
     "/Filter": "/FlateDecode",
     "/FormType": 1, 
     "/Length": 5178, 
     "/PTEX.FileName": "./ntg.pdf","/PTEX.InfoDict": "13 0 R","/PTEX.PageNumber": 1, 
     "/Resources": {
          "/Font": {"/F10": "34 0 R",…,"/F8": "37 0 R"
     },
     "/ProcSet": ["/PDF","/Text"]},
     "/Subtype": "/Form","/Type": "/XObject"
},
```
My code to process this uses QPDFPageObjectHelper.  I get the page-QPDFObjectHandle fine, as well as the resources-QPDFObjectHandle ([8 0 R]).  I find the /XObject in the resource dictionary, so I'm at the "{"/Im4": "7 0 R"}".  

I get {"/Im4": "7 0 R"} as a dictionary. When I use the "/Im4" key to get the "7 0 R" object - I'm expecting the "7 0 R" object to be a dictionary, but it's a stream.  When I stream it using `pipeStreamData`, it's actually the text for page 1.

So: I can't find the "7 0 R" dictionary (/BBox, /Filter, ... /Subtype=/Form, /Type=/XObject) that `qpdf --json` finds.

What am I doing wrong?  If needed, I can try and post a mini-working example, but it would take a little time, so I thought I'd just see first if I was missing something simple.

 (I'm using QPDF 10.0.1.)

**Comment by jberkenbilt** [comment]
Form XObjects are streams. They are like pages in many ways, but unlike pages, they are streams, and the bodies of the streams are the contents. QPDF 10.0.0 introduced the the `filterAsContents` method that allows you to filter the contents of form XObjects like page content streams, but otherwise, qpdf doesn't have very many helper APIs for dealing with form XObjects. I actually have this on the TODO list for the next qpdf release. In particular though, QPDFPageObjectHelper doesn't work all that well with form XObjects in the current release.

To get a dictionary for a stream, you use getDict() on the stream. So once you resolve 7 0 R as a stream, use getDict to find the keys. Then you can use pipStreamData on 7 0 R itself.

Let me know if this helps. I'm happy to answer additional questions and to better understand your use case. It might help me create better APIs for form XObjects.

**Comment by drcdr** [comment]
Thanks @jberkenbilt for the clarification.  I'd sort-of tried that, but my problem is I'd used my own version of getDictAsMap() that only worked only on dictionaries, I hadn't used the stream getDict() yet.  (For debugging, I use my own getDictAsMap that returns a pair of the QPDFObjectHandle and it's unparse(), so I can see what's going on in the debugger.  This didn't work for a stream, without doing getDict() as you mentioned first.)  Fortunately, qpdf.cc has so much in it, to show how to do this.  I'd missed the part of using getDict() on the stream, there.

To address the use case: what I'm trying to do is similar to what is being done [here](https://github.com/qpdf/qpdf/blob/master/examples/pdf-parse-content.cc#L90-L92), but just extending the processing for when the page has form XObjects; maybe something like this, for now:

```
std::vector<QPDFPageObjectHelper> pages = QPDFPageDocumentHelper(pdf).getAllPages();
QPDFPageObjectHelper& page = pages.at(QIntC::to_size(pageno-1));
ParserCallbacks cb;
page.parsePageContents(&cb);

QPDFObjectHandle stream;
for (auto& stream : pages_fxo.at(pageno-1).streams)  // could be more than 1
     stream.parseContentStream(stream, callbacks);
```

**Comment by jberkenbilt** [comment]
Yeah, this is probably a good start. I'll leave this open and mark it as an enhancement. It would be nice if there were something in the library to get pages_fxo, possibly recursively. I've written that code twice now, once in the library itself and once in an external program I wrote for a different purpose. You may also want to keep in mind that form XObjects can contain other form XObjects. I'll try to make this cleaner in a future version, but calling parseContentStream on the form XObject as you have done here should work fine. If it doesn't, let me know.

**Comment by jberkenbilt** [comment]
qpdf 10.1 will have several new methods in `QPDFPageObjectHelper` to help work with form XObjects that handle recursing into nested form XObjects, and several of the existing methods in that class will work with form XObjects as well as pages. The commit that closes this ticket adds some forEach method, but the enhancements are spread out over several commits and are detailed in ChangeLog. As I write this comment, I have not yet completed the enhancements, so this issue should be considered fully resolved when 10.1 is released.
